pkg/alertmanager/state_replication.go

```diff
@@ -0,0 +1,193 @@
+func (s *state) AddState(key string, cs cluster.State, _ prometheus.Registerer) cluster.ClusterChannel {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.states[key] = cs
+
+	s.partialStateMergesTotal.WithLabelValues(key)
+	s.partialStateMergesFailed.WithLabelValues(key)
+	s.stateReplicationTotal.WithLabelValues(key)
+	s.stateReplicationFailed.WithLabelValues(key)
+
+	send := func(b []byte) {
+		s.msgc <- &clusterpb.Part{Key: key, Data: b}
+	}
+
+	return newStateChannel(key, send, s.logger)
+}
+
+// MergePartialState merges a received partial message with an internal state.
+func (s *state) MergePartialState(p *clusterpb.Part) error {
+	s.partialStateMergesTotal.WithLabelValues(p.Key).Inc()
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	st, ok := s.states[p.Key]
+	if !ok {
+		level.Error(s.logger).Log("msg", "state key not found while merging", "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return fmt.Errorf("unable to merge state, key '%s' not found while merging", p.Key)
+	}
+
+	if err := st.Merge(p.Data); err != nil {
+		level.Error(s.logger).Log("msg", "failed to merge state", "err", err, "key", p.Key)
+		s.partialStateMergesFailed.WithLabelValues(p.Key).Inc()
+		return errors.Wrap(err, "unable to merge state")
+	}
+
+	return nil
+}
+
+// The clusterWait function uses the position to determine how long should it wait before sending a notification.
+// In the Cortex case, we copy the function directly and use the manager to figure out which position we hold.
+func (s *state) Position() int { return 0 }
+
+// Settle waits until the alertmanagers are ready (and sets the appropriate internal state when it is).
+// The idea is that we don't want to start working" before we get a chance to know most of the notifications and/or silences.
+func (s *state) Settle(ctx context.Context, _ time.Duration) {
+	level.Info(s.logger).Log("msg", "Waiting for notification and silences to settle...")
+
+	// TODO: Make sure that the state is fully synchronised at this point.
+	// We can check other alertmanager(s) and explicitly ask them to propagate their state to us if available.
+	close(s.readyc)
+}
+
+// WaitReady is needed for the pipeline builder to know whenever we've settled and the state is up to date.
+func (s *state) WaitReady() {
+	//TODO: At the moment, we settle in a separate go-routine (see multitenant.go as we create the Peer) we should
+	// mimic that behaviour here once we have full state replication.
+	s.Settle(context.Background(), time.Second)
+	<-s.readyc
+}
+
+func (s *state) Ready() bool {
+	select {
+	case <-s.readyc:
+		return true
+	default:
+	}
+	return false
+}
+
+func (s *state) handleBroadcasts() {
+	for {
+		select {
+		case p := <-s.msgc:
+			// If the replication factor is <= 1, we don't need to replicate any state anywhere else.
+			if s.replicationFactor <= 1 {
+				return
+			}
+
+			s.stateReplicationTotal.WithLabelValues(p.Key).Inc()
+			ctx := context.Background() //TODO: I probably need a better context
```

Comment:
Hint: If `state` was a `services.Service` (= object with lifecycle in Cortex), it would already have its own context. :)

Reply:
Can you help me break this down? 

I have not seen us use services within services (That's different from services with sub-services) and even if we do, at its current state it seems like overkill. It's more code to manage the service, in particular, because the State itself it's satisfied by two different things (our `state` or the upstream `Peer`) which in my eyes brings us little benefits.

Reply:
I'm only talking about `state` being a service (and not `Peer`), since it runs background process and its lifecycle needs to be managed.

Another possibility is to use custom context initialized in new, and canceled when stopping.